File: src/main/java/com/pharmacy/Controllers/UserController.java
package com.pharmacy.Controllers;

import com.pharmacy.Database.DatabaseInstance;
import com.pharmacy.Models.UserModel;
import com.pharmacy.Utils.EncryptionUtils;
import com.pharmacy.Views.UsersPage;

import javax.swing.JOptionPane;
import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.sql.Statement;

public class UserController {

    Connection conn = null;
    PreparedStatement prepStatement = null;
    Statement statement = null;
    ResultSet resultSet = null;

    public UserController() {
        try {
            conn = new DatabaseInstance().getConnection();
            statement = conn.createStatement();
        } catch (SQLException ex) {
            ex.printStackTrace();
        }
    }

    // Methods to add new user
    public void addUser(UserModel userModel) {
        try {
            String duplicateQuery = "SELECT * FROM users WHERE name='"
                + userModel.getName()
                + "' AND phone='"
                + userModel.getPhone()
                + "' AND user_type='"
                + userModel.getType()
                + "'";
            resultSet = statement.executeQuery(duplicateQuery);

            if (resultSet.next()) {
                JOptionPane.showMessageDialog(null, "User already exists");
            } else {
                String encryptedPass = new EncryptionUtils().encrypt(userModel.getPassword());

                String insertQuery = "INSERT INTO users (name,phone,username,password,user_type) "
                    + "VALUES(?,?,?,?,?)";
                prepStatement = conn.prepareStatement(insertQuery);
                prepStatement.setString(1, userModel.getName());
                prepStatement.setString(2, userModel.getPhone());
                prepStatement.setString(3, userModel.getUsername());
                prepStatement.setString(4, encryptedPass);
                prepStatement.setString(5, userModel.getType());

                prepStatement.executeUpdate();
            }
        } catch (Exception ex) {
            ex.printStackTrace();
        }
    }

    /**
     * Update existing user (password excluded!)
     * <p>
     * See {@link #updatePass(int id, String password)} for updating
     * password
     *
     * @param userModel populated UserModel
     */
    public void updateUser(UserModel userModel) {

        try {
            String query = "UPDATE users SET username=?,name=?,phone=?,user_type=? WHERE id=?";
            prepStatement = conn.prepareStatement(query);
            prepStatement.setString(1, userModel.getUsername());
            prepStatement.setString(2, userModel.getName());
            prepStatement.setString(3, userModel.getPhone());
            prepStatement.setString(4, userModel.getType());
            prepStatement.setInt(5, userModel.getId());

            prepStatement.executeUpdate();
        } catch (SQLException throwables) {
            throwables.printStackTrace();
        }
    }

    // Method to delete existing user
    public void deleteUser(String id) {
        try {
            String query = "DELETE FROM users WHERE id=?";
            prepStatement = conn.prepareStatement(query);
            prepStatement.setString(1, id);

            prepStatement.executeUpdate();
        } catch (SQLException sqlException) {
            sqlException.printStackTrace();
        }
        new UsersPage().loadDataSet();
    }

    // Method to retrieve data set to display in table
    public ResultSet getUsers() {
        try {
            String query = "SELECT id, name, username, phone, user_type FROM users";
            resultSet = statement.executeQuery(query);
        } catch (SQLException sqlException) {
            sqlException.printStackTrace();
        }

        return resultSet;
    }

    public ResultSet searchUsers(String search) {
        try {
            String query = "SELECT * FROM users WHERE name LIKE '%" + search + "%' OR phone LIKE '%" + search + "%' OR user_type LIKE '%" + search + "%'";
            resultSet = statement.executeQuery(query);
        } catch (SQLException sqlException) {
            sqlException.printStackTrace();
        }

        return resultSet;
    }

    public ResultSet findUser(String username) {
        try {
            String query = "SELECT id, username, name, phone, user_type FROM users WHERE username='" + username + "'";
            resultSet = statement.executeQuery(query);
        } catch (SQLException sqlException) {
            sqlException.printStackTrace();
        }

        return resultSet;
    }

    public ResultSet getUserLogs() {
        try {
            String query = "SELECT users.name,userlogs.username,in_time,out_time FROM userlogs"
                + " INNER JOIN users on userlogs.username=users.username ORDER BY in_time DESC";
            resultSet = statement.executeQuery(query);
        } catch (SQLException sqlException) {
            sqlException.printStackTrace();
        }

        return resultSet;
    }

    public void addUserLogin(UserModel userModel) {
        try {
            String query = "INSERT INTO userlogs (username, in_time, out_time) values(?,?,?)";
            prepStatement = conn.prepareStatement(query);
            prepStatement.setString(1, userModel.getUsername());
            prepStatement.setString(2, userModel.getInTime());
            prepStatement.setString(3, userModel.getOutTime());

            prepStatement.executeUpdate();
        } catch (SQLException e) {
            e.printStackTrace();
        }
    }

    public boolean matchPasswords(String username, String password) {
        try {
            String query = "SELECT password FROM users WHERE username='"
                + username
                + "'";
            resultSet = statement.executeQuery(query);

            if (resultSet.next()) {
                String decryptedPass = new EncryptionUtils().decrypt(resultSet.getString("password"));

                return decryptedPass.equals(password);
            }
        } catch (SQLException sqlException) {
            sqlException.printStackTrace();
        }

        return false;
    }

    public void updatePass(int id, String password) {
        try {
            String encryptedPass = new EncryptionUtils().encrypt(password);

            String query = "UPDATE users SET password=? WHERE id='" + id + "'";
            prepStatement = conn.prepareStatement(query);
            prepStatement.setString(1, encryptedPass);

            prepStatement.executeUpdate();
        } catch (SQLException sqlException) {
            sqlException.printStackTrace();
        }
    }
}

File: src/main/java/com/pharmacy/controllers/UserController.kt
package com.pharmacy.controllers

import com.pharmacy.database.DatabaseInstance
import com.pharmacy.models.UserModel
import com.pharmacy.utils.EncryptionUtils
import java.sql.Connection
import java.sql.PreparedStatement
import java.sql.ResultSet
import java.sql.SQLException
import java.sql.Statement
import javax.swing.JOptionPane

class UserController(private var logId: Int) {
    private var connection: Connection? = null
    private var statement: Statement? = null

    init {
        try {
            connection = DatabaseInstance().getConnection()
            statement = connection!!.createStatement()
        } catch (e: SQLException) {
            throw SQLException(e)
        }
    }

    val users: ResultSet?
        get() {
            val resultSet: ResultSet?

            try {
                val query = "SELECT id, name, username, phone, user_type FROM users"
                resultSet = statement!!.executeQuery(query)
            } catch (e: SQLException) {
                throw SQLException(e)
            }

            return resultSet
        }

    // Methods to add new user
    fun addUser(userModel: UserModel) {
        val resultSet: ResultSet?

        try {
            val duplicateQuery = """
                SELECT * FROM users
                WHERE name=?
                AND phone=?
                AND user_type=?
                """.trimIndent()

            val duplicateStatement = connection!!.prepareStatement(duplicateQuery)
            duplicateStatement.setString(1, userModel.name)
            duplicateStatement.setString(2, userModel.phone)
            duplicateStatement.setString(3, userModel.type)

            resultSet = duplicateStatement.executeQuery()
            if (resultSet!!.next()) {
                JOptionPane.showMessageDialog(null, "User already exists")
            } else {
                val encryptionUtils = EncryptionUtils()
                val secretKey = encryptionUtils.generateKeyBytes()

                val insertQuery = """
                    INSERT INTO users (name,phone,username,password,user_type,secret_key)
                    VALUES(?,?,?,?,?,?);
                    """.trimIndent()

                val prepStatement: PreparedStatement = connection!!.prepareStatement(insertQuery)
                prepStatement.setString(1, userModel.name)
                prepStatement.setString(2, userModel.phone)
                prepStatement.setString(3, userModel.username)
                prepStatement.setBytes(4, encryptionUtils.encrypt(userModel.password!!, secretKey))
                prepStatement.setString(5, userModel.type)
                prepStatement.setBytes(6, secretKey)

                prepStatement.executeUpdate()
                LogsController().addLogEntry(
                    logId,
                    "Added new user: ${userModel.name} (${userModel.username})"
                )
            }
        } catch (e: SQLException) {
            throw SQLException(e)
        }
    }

    /**
     * Update existing user (password excluded!)
     * See [.updatePass] for updating password
     *
     * @param userModel populated UserModel
     */
    fun updateUser(userModel: UserModel) {
        try {
            val query = "UPDATE users SET username=?,name=?,phone=?,user_type=? WHERE id=?"
            val prepStatement: PreparedStatement = connection!!.prepareStatement(query)

            prepStatement.setString(1, userModel.username)
            prepStatement.setString(2, userModel.name)
            prepStatement.setString(3, userModel.phone)
            prepStatement.setString(4, userModel.type)
            prepStatement.setInt(5, userModel.id!!)

            prepStatement.executeUpdate()
            LogsController().addLogEntry(
                logId,
                "Updated user: ${userModel.name} (${userModel.username})"
            )
        } catch (e: SQLException) {
            throw SQLException(e)
        }
    }

    // Method to delete existing user
    fun deleteUser(id: Int, username: String) {
        try {
            val query = "DELETE FROM users WHERE id=?"

            val prepStatement: PreparedStatement = connection!!.prepareStatement(query)
            prepStatement.setInt(1, id)

            prepStatement.executeUpdate()
            LogsController().addLogEntry(logId, "Deleted user: $username")
        } catch (e: SQLException) {
            throw SQLException(e)
        }
    }


    fun searchUsers(search: String): ResultSet? {
        val resultSet: ResultSet?

        try {
            val query = """
                SELECT id, name, username, phone, user_type
                FROM users
                WHERE name LIKE ?
                OR username LIKE ?
                OR phone LIKE ?
                OR user_type LIKE ?;
                """.trimIndent()

            val prepStatement: PreparedStatement = connection!!.prepareStatement(query)
            prepStatement.setString(1, "%$search%")
            prepStatement.setString(2, "%$search%")
            prepStatement.setString(3, "%$search%")
            prepStatement.setString(4, "%$search%")

            resultSet = prepStatement.executeQuery()
        } catch (e: SQLException) {
            throw SQLException(e)
        }

        return resultSet
    }

    fun getUserId(username: String): Int {
        val resultSet: ResultSet?
        var id = 0

        try {
            val q = """
                SELECT id
                FROM users
                WHERE username=?
                """.trimIndent()

            val prepStatement: PreparedStatement = connection!!.prepareStatement(q)
            prepStatement.setString(1, username)

            resultSet = prepStatement.executeQuery()
            if (resultSet!!.next()) {
                id = resultSet.getInt("id")
            }
        } catch (e: SQLException) {
            throw SQLException(e)
        }

        return id
    }

    fun findUser(username: String): ResultSet? {
        val resultSet: ResultSet?

        try {
            val q = """
                SELECT id, username, name, phone, user_type
                FROM users
                WHERE username=?
                """.trimIndent()

            val prepStatement: PreparedStatement = connection!!.prepareStatement(q)
            prepStatement.setString(1, username)

            resultSet = prepStatement.executeQuery()
        } catch (e: SQLException) {
            throw SQLException(e)
        }

        return resultSet
    }

    fun matchPasswords(username: String, password: String): Boolean {
        val resultSet: ResultSet?

        try {
            val q = """
                SELECT password, secret_key
                FROM users
                WHERE username=?
                """.trimIndent()

            val prepStatement: PreparedStatement = connection!!.prepareStatement(q)
            prepStatement.setString(1, username)

            resultSet = prepStatement.executeQuery()
            if (resultSet!!.next()) {
                val secretKey = resultSet.getBytes("secret_key")
                val encryptedPassword = resultSet.getBytes("password")

                val decryptedPass = EncryptionUtils().decrypt(encryptedPassword, secretKey)
                return decryptedPass == password
            }
        } catch (e: SQLException) {
            throw SQLException(e)
        }

        return false
    }

    fun updatePass(id: Int, username: String, password: String?) {
        try {
            // encryption
            val encryptionUtils = EncryptionUtils()
            val secretKey = encryptionUtils.generateKeyBytes()
            val encryptedPass = encryptionUtils.encrypt(password!!, secretKey)

            // db query
            val query = "UPDATE users SET password=?, secret_key=? WHERE id=?"

            val prepStatement: PreparedStatement = connection!!.prepareStatement(query)
            prepStatement.setBytes(1, encryptedPass)
            prepStatement.setBytes(2, secretKey)
            prepStatement.setInt(3, id)

            prepStatement.executeUpdate()
            LogsController().addLogEntry(logId, "Password updated for user: $username")
        } catch (e: SQLException) {
            throw SQLException(e)
        }
    }
}
